be/src/exec/workgroup/pipeline_executor_set_manager.cpp

```diff
@@ -0,0 +1,207 @@
+    if (const Status status = executors->start(); !status.ok()) {
+        LOG(WARNING) << "[WORKGROUP] failed to start executors for workgroup "
+                     << "[workgroup=" << wg->to_string() << "] "
+                     << "[conf=" << _conf.to_string() << "] "
+                     << "[cpuids=" << CpuUtil::to_string(cpuids) << "] "
+                     << "[status=" << status << "]";
+        LOG(INFO) << "[WORKGROUP] assign shared executors to workgroup "
+                  << "[workgroup=" << wg->to_string() << "] ";
+        executors->close();
+        wg->set_executors(_shared_executors.get());
+        return _shared_executors.get();
+    }
+
+    LOG(INFO) << "[WORKGROUP] assign dedicated executors to workgroup "
+              << "[workgroup=" << wg->to_string() << "] ";
+    wg->set_exclusive_executors(std::move(executors));
+    return wg->executors();
+}
+
+void ExecutorsManager::change_num_connector_scan_threads(uint32_t num_connector_scan_threads) {
+    const auto prev_val = _conf.num_total_connector_scan_threads;
+    _conf.num_total_connector_scan_threads = num_connector_scan_threads;
+    if (_conf.num_total_connector_scan_threads == prev_val) {
+        return;
+    }
+
+    for_each_executors([](const auto& executors) { executors.notify_num_total_connector_scan_threads_changed(); });
+}
+
+void ExecutorsManager::change_enable_resource_group_cpu_borrowing(bool val) {
+    if (_conf.enable_cpu_borrowing == val) {
+        return;
+    }
+    _conf.enable_cpu_borrowing = val;
+
+    update_shared_executors();
+}
+
+void ExecutorsManager::for_each_executors(const ExecutorsConsumer& consumer) const {
+    for (const auto& [_, wg] : _parent->_workgroups) {
+        if (wg != nullptr && wg->exclusive_executors() != nullptr) {
+            consumer(*wg->exclusive_executors());
+        }
+    }
+    if (_shared_executors) {
+        consumer(*_shared_executors);
+    }
+}
+
+bool ExecutorsManager::should_yield(const WorkGroup* wg) const {
+    if (!_conf.enable_cpu_borrowing) {
+        return false;
+    }
+
+    // Only resource groups without dedicated executors can borrow CPU.
+    if (wg->exclusive_executors() != nullptr) {
+        return false;
+    }
+
+    const Thread* thread = Thread::current_thread();
+    if (thread == nullptr) {
+        return false;
+    }
+    const auto it = _cpu_owners.find(thread->first_bound_cpuid());
+    if (it == _cpu_owners.end()) {
+        return false;
+    }
+
+    // Check before using `get_wg`, which is a little heavy.
+    if (it->second.raw_wg == nullptr) {
+        return false;
+    }
+    const auto owner_wg = it->second.get_wg();
+    return owner_wg != nullptr && owner_wg->num_running_queries() > 0;
+}
+
+void ExecutorsManager::CpuOwnerContext::set_wg(WorkGroup* new_wg) {
+    // TODO: use std::atomic<std::shared_ptr> instead of std::shared_ptr with atomic load/store,
+    // when our compiler support it.
```

Comment:
this will be supported if we update compiler to gcc 14?

Reply:
Yes, we could modify this after upgrading to gcc 14.